### ISSUE TITLE
RFC-2: Mixing base16 and base64

### ISSUE THREAD

**Issue by adrianhopebailie** [interledger/rfcs#78]
The string encoding of a condition is:
`"cc:" BASE16(type) ":" BASE16(featureBitmask) ":" BASE64URL(fingerprint) ":" BASE10(maxFulfillmentLength)`

and of a fulfillment:

`"cf:" BASE16(type) ":" BASE64URL(payload)`

We are taking raw bytes and string encoding them using two different encoding formats. This is unnecessarily confusing.

The string encoding is not going to be used if efficiency is required so we should simply use base16 and show all binary data as hex.

If you require efficiency and want to transmit a string use the binary encoding and then re-encode using base64.

**Comment by adrianhopebailie** [comment]
@justmoon @emschwartz any thoughts on this?

It took me 20 minutes to figure out all the encoding and decoding steps in the js crypto-conditions unit tests which is what prompted this question.

Wouldn't it be easier to just use hex for the string encoding? 
- Less error prone
- Encoding and decoding supported by default in most languages.
- Consistent across all fields

**PROPOSAL**

Use BASE10 for all numbers or just BASE16 for everything. "type" is a numerically indexed enumeration, not vastly different to "maxFulfillmentLength".

- `"cc:" BASE16(type) ":" BASE16(featureBitmask) ":" BASE16(fingerprint) ":" BASE16(maxFulfillmentLength)`
- `"cf:" BASE16(type) ":" BASE16(payload)`

**Comment by justmoon** [comment]
I don't like `BASE16(maxFulfillmentLength)`, because it's very unreadable. How many bytes is `100000` (hex)? Answer: It's 1MB, but would you have known that? 1048576 on the other hand is a million so clearly around 1MB.

For the fulfillment, I don't like `BASE16(payload)`, because hex encoding/decoding is quite a bit slower than base64.

I've been thinking that we should treat conditions and fulfillments more differently since the requirements are so different.

Maybe fulfillments should just be the base64-encoded binary fulfillment, i.e. same format as the binary version, just base64 encoded and (maybe) prefixed with `cf:`. That way we get rid of one of the formats.

On the condition side - I'd like to get rid of the bitmask for simple condition types (ones without subconditions), so we could move it into the fingerprint for those conditions that need it.

The type I'm pretty sure is already base-10. If not, I'm totally onboard with changing it to that.

That leaves us with:

`"cc:" BASE10(type) ":" BASE64(fingerprint) ":" BASE10(maxFulfillmentLength)`
`"cf:" BASE64(fulfillment)`

**Comment by jtremback** [comment]
> On the condition side - I'd like to get rid of the bitmask for simple condition types (ones without subconditions), so we could move it into the fingerprint for those conditions that need it.

This kind of stuff is the exact reason that I don't support the current formalization of string encodings at all. This is the bikiest of sheds, and the string format isn't even readable or useful anyway.

Let's get this standard working and adopted and then release another RFC with an optional string encoding.

**Comment by adrianhopebailie** [comment]
> the string format isn't really that readable or useful anyway (IMO)

I tend to agree and that is why I really don't like base64. It has a too many variants and peculiarities for little benefit.

The string encoding should be verbose and easy to read. As soon as we start combining fields then we lose any value in having a string encoding at all.

For most OO languages we'll likely only care about the binary format anyway because once it's been parsed then you'll have a nice language specific object representation.

> hex encoding/decoding is quite a bit slower than base64

I am surprised to hear this. Most languages/platforms have a `toString` or equivalent function that will convert any number or byte array to a hex string and my assumption is that these are highly optimized.

The fact that we have to ship base64 utils with the lib suggests this is "non-standard" 

**JSON?**

This may sound like an odd suggestion but perhaps we should have binary and JSON encoding? Then it makes sense for the binary fields in the JSON encoded data to be base64url encoded.

The current string encoding is really URI encoding. This occurred to me when was trying to decide on the data type of the String encoded form in Java and couldn't decide between a `java.net.URI` and `java.lang.String`.

The benefit of the former is some format validation up-front (although arguably not much) and any other value you get from a condition/fulfillment being a valid URI.

If you are looking for a nice human readable form then JSON is way better.

So you could have 3 encoding formats:

1. Binary (OER as defined) - most common as it's the most efficient
2. URI (as I proposed above) - if you want to represent a condition/fulfillment as a URI? 
3. JSON

```json
{
  "type" : "preimageSha256",
  "features" : ["sha256", "preimage"],
  "fingerprint" : "fytfweu486sduf2kyia65difywekutfwedo6woef3o6f",
  "maxfulfillment_length" : 256
}
```

**Comment by jtremback** [comment]
🚴🏡

**Comment by adrianhopebailie** [comment]
@jtremback - we're defining a protocol and this is the core. Not easy to change later.

If all you care about is programming with conditions then just use the binary encoding and forget about the others.

**Comment by jtremback** [comment]
Apologies for my tone. The way I see it, is that the binary encoding is necessarily core, because it needs to be used for subconditions and subfulfillments. It also needs to be used as input to any hashing (please correct me if I'm wrong on any of this).

The string encoding just seems like something that's there to make the top level condition prettier. There's nothing wrong with this, but I think it's better defined in a separate RFC. I don't actually feel that strongly about it though, so I'll leave it to yall.

**Comment by adrianhopebailie** [comment]
> Apologies for my tone

Not at all. I agree we're bike-shedding but as you say it's not a blocker. The binary encoding is what we need for complex conditions.

BUT, the string encoding is being used in many of the other pieces of work so we do need to figure it out for the sake of those.

Re: whether or not it should be in a separate RFC... I'm not fussed either way

**Comment by adrianhopebailie** [comment]
Addressed in #92

**Comment by adrianhopebailie** [comment]
Addressed in #92
